crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl

```diff
@@ -0,0 +1,171 @@
+    // WebGL2 structs must be 16 byte aligned.
+    _webgl2_padding: vec3<f32>
+#endif
+}
+@group(0) @binding(4) var<uniform> settings: MotionBlur;
+@group(0) @binding(5) var<uniform> globals: Globals;
+
+@fragment
+fn fragment(
+    #ifdef MULTISAMPLED
+        @builtin(sample_index) sample_index: u32,
+    #endif
+    in: FullscreenVertexOutput
+) -> @location(0) vec4<f32> { 
+    let texture_size = vec2<f32>(textureDimensions(screen_texture));
+    let frag_coords = vec2<i32>(in.uv * texture_size);
+
+#ifdef MULTISAMPLED
+    let base_color = textureLoad(screen_texture, frag_coords, i32(sample_index));
+#else
+    let base_color = textureSample(screen_texture, texture_sampler, in.uv);
+#endif
+
+    if i32(settings.samples) <= 0 || settings.shutter_angle <= 0.0 {
+        return base_color;
+    }
+
+    let shutter_angle = settings.shutter_angle;
+
+#ifdef MULTISAMPLED
+    let this_motion_vector = textureLoad(motion_vectors, frag_coords, i32(sample_index)).rg;
+#else
+    let this_motion_vector = textureSample(motion_vectors, texture_sampler, in.uv).rg;
+#endif
+
+#ifdef NO_DEPTH_TEXTURE_SUPPORT
+    let this_depth = 0.0;
+    let depth_supported = false;
+#else
+    let depth_supported = true;
+#ifdef MULTISAMPLED
+    let this_depth = textureLoad(depth, frag_coords, i32(sample_index));
+#else
+    let this_depth = textureSample(depth, texture_sampler, in.uv);
+#endif
+#endif
+    
+    // The exposure vector is the distance that this fragment moved while the camera shutter was
+    // open. This is the motion vector (total distance traveled) multiplied by the shutter angle (a
+    // fraction). In film, the shutter angle is commonly 0.5 or "180 degrees" (out of 360 total).
+    // This means that for a frame time of 20ms, the shutter is only open for 10ms.
+    //
+    // Using a shutter angle larger than 1.0 is non-physical, objects would need to move further
+    // than they physically travelled during a frame, which is not possible.
+    let exposure_vector = shutter_angle * this_motion_vector;
+
+    var accumulator: vec4<f32>;
+    var weight_total = 0.0;
+    let n_samples = i32(settings.samples);
+    let noise = hash_noise(frag_coords, globals.frame_count); // 0 to 1
+       
+    for (var i = -n_samples; i < n_samples; i++) {
+        // The current sample step vector, from in.uv
+        let step_vector = 0.5 * exposure_vector * (f32(i) + noise) / f32(n_samples);
+        var sample_uv = in.uv + step_vector;
+        let sample_coords = vec2<i32>(sample_uv * texture_size);
+
+    #ifdef MULTISAMPLED
+        let sample_color = textureLoad(screen_texture, sample_coords, i32(sample_index));
+    #else
+        let sample_color = textureSample(screen_texture, texture_sampler, sample_uv);
+    #endif
+    #ifdef MULTISAMPLED
+        let sample_motion = textureLoad(motion_vectors, sample_coords, i32(sample_index)).rg;
+    #else
+        let sample_motion = textureSample(motion_vectors, texture_sampler, sample_uv).rg;
```

Comment:
Do you need to check that motion vectors don't point off screen?

Reply:
no, they are only used to check motion similarity